File: 3-hard/00051-n-queens/solution.py
class Solution:
    def is_safe(self, x: int, y: int, max: int, board):
        for ofs in range(0,max): 
            if x  >= ofs and y >= ofs and board[x-ofs][y-ofs]:
                return False
            if x + ofs < max and y >= ofs and board[x + ofs][y - ofs]:
                return False
            if y >= ofs and board[x][y - ofs]:
                return False
        return True

    def add_board_state_into_res(self, max_row, board, res):
        v = []
        for y in range(max_row):
            tmp = ""
            for x in range(max_row):
                if board[x][y]:
                    v.append("." * x + "Q" + "." * (max_row - x -1))
        res.append(v)

    def solve_helper(self, current_row, max_row,  board, res):
        for x in range(max_row):
            if self.is_safe(x, current_row, max_row, board):
                board[x][current_row] = True
                if current_row == max_row - 1:
                    self.add_board_state_into_res(max_row, board, res)
                else:
                    self.solve_helper(current_row + 1, max_row, board, res)
                board[x][current_row] = False

    def solveNQueens(self, n: int) -> List[List[str]]:
        board = [[False for _ in range(n)] for _ in range(n)]
        res = []
        self.solve_helper(0, n, board, res)
        return res

File: 3-hard/00051-n-queens/solution.rs
impl Solution {
    fn is_safe(x: usize, y: usize, max: usize, board: &mut Vec<Vec<bool>>) -> bool {
        for ofs in 0..max {
            if x  >= ofs && y >= ofs && board[x-ofs][y-ofs] {
                return false;
            }
            if x + ofs < max && y >= ofs && board[x + ofs][y - ofs] {
                return false;
            }
            if y >= ofs && board[x][y - ofs] {
                return false;
            }
        }
        return true;
    }
    fn add_board_state_into_res(max: usize, board: &mut Vec<Vec<bool>>, res: &mut Vec<Vec<String>>) {
        let mut v: Vec<String> = Vec::new();
        for y in 0..max {
            let mut tmp: String = String::new();
            for x in 0..max {
                if(board[x][y]) {                         
                    tmp.push('Q');
                } else {
                    tmp.push('.');
                }
            }
            v.push(tmp);
        }
        res.push(v);                   
    }
    fn solve_helper(current_row: usize, max: usize,  board: &mut Vec<Vec<bool>>, res: &mut Vec<Vec<String>>) {
        for x in 0..max {
            if Self::is_safe(x, current_row, max, board) {
                board[x][current_row] = true;
                if current_row == max-1 {
                    Self::add_board_state_into_res(max, board, res);
                } else {
                    Self::solve_helper(current_row + 1, max, board, res);
                }
                board[x][current_row] = false;
            }
        }      
    }
    pub fn solve_n_queens(n: i32) -> Vec<Vec<String>> {
        let mut board = vec![vec![false;n as usize];n as usize];
        let mut res: Vec<Vec<String>> = Vec::new();
        Self::solve_helper(0usize, n as usize, &mut board, &mut res);
        res
    }
}
